src/patchset.py

```diff
@@ -0,0 +1,250 @@
+    def _apply_patches(self, kdir, patch_artifacts):
+        patchset_hash = hashlib.sha256()
+        for patch_name, patch_url in patch_artifacts.items():
+            self.log.info(
+                f"Applying patch {patch_name}, url: {patch_url}"
+            )
+            patch_hash = self._apply_patch(kdir, patch_url)
+            patchset_hash.update(patch_hash.encode("utf-8"))
+
+        return patchset_hash.hexdigest()
+
+    def _download_checkout_archive(self, tarball_url, retries=3):
+        self.log.info(f"Downloading checkout tarball, url: {tarball_url}")
+        urlpath = urlparse(tarball_url).path
+        tar_filename = os.path.basename(urlpath)
+        kernelci.build.pull_tarball(
+            kdir=self._kdir,
+            url=tarball_url,
+            dest_filename=tar_filename,
+            retries=retries,
+            delete=True
+        )
+
+    def _update_node(
+        self,
+        patchset_node,
+        revision,
+        tarball_url,
+    ):
+        node = patchset_node.copy()
+        node.update({
+            'revision': revision,
+            'state': 'available',
+            'artifacts': {
+                'tarball': tarball_url,
+            },
+            'holdoff': str(datetime.utcnow() + timedelta(minutes=10))
+        })
+        try:
+            self._api.update_node(node)
+        except requests.exceptions.HTTPError as err:
+            err_msg = json.loads(err.response.content).get("detail", [])
+            self.log.error(err_msg)
+
+    def _setup(self, *args):
+        return self._api_helper.subscribe_filters({
+            'op': 'created',
+            'name': 'patchset',
+            'state': 'running',
+        })
+
+    def _has_approved_domain(self, url):
+        return urlparse(url).hostname in ALLOWED_DOMAINS
+
+    def _process_patchset(self, checkout_node, patchset_node):
+        build_config = self._find_build_config(checkout_node)
+        if not build_config:
+            raise RuntimeError(
+                "No build config found for checkout node %s",
+                checkout_node['id']
+            )
+
+        patch_artifacts = patchset_node.get("artifacts")
+        if not patch_artifacts:
+            raise ValueError(
+                f"No patch artifacts available for node {patchset_node['id']}"
+            )
+
+        if not all(
+            self._has_approved_domain(patch_mbox_url)
+            for patch_mbox_url in patch_artifacts.values()
+        ):
+            raise RuntimeError("Forbidden domain")
+
+        patchset_build_config = self._find_build_config(patchset_node)
+        if build_config != patchset_build_config:
+            raise ValueError(
+                f"Patchsets node {patchset_node['id']} build config "
+                f"doesn't match to the parent node {checkout_node['id']}"
+            )
+
+        self._download_checkout_archive(checkout_node["artifacts"]["tarball"])
+
+        chekout_name = '-'.join([
```

Comment:
About the patch format, this would ideally need to be clearly defined in a written form as far as what is expected from the API point of view when submitting a patchset.  It matters because it means the checksum will be different depending on what is being included.  I think the ideal approach is to require having just the commit subject, message and diff of course.  Additional meta-data e.g. commit author, date/time, Git notes etc. shouldn't be required.  This is to enable collating results together from a same patchset applied in different conditions.

Using `patch` to apply the diff is fine as that's the least common denominator, I don't think it's needed to recreate a local git history.  So regardless of what's in the patch files it should work and produce the same build.  Also, the initial kernel source tree comes from a tarball with no git history anyway.